beacon_node/network/src/beacon_processor/work_reprocessing_queue.rs

```diff
@@ -665,6 +691,24 @@ impl<T: BeaconChainTypes> ReprocessQueue<T> {
                     }
                 }
             }
+            InboundEvent::Msg(BackfillSync(work_event)) => {
+                let position_in_queue = self.backfill_delay_queue.len();
+                let slot_duration = slot_clock.slot_duration();
+
+                let initial_delay = slot_clock
+                    .duration_to_next_slot()
+                    .map_or(slot_duration, |duration_to_next_slot| {
+                        duration_to_next_slot + (slot_duration / 2)
+                    });
+
+                // FIXME(jimmy): we should try to process multiple batches per slot
+                let scheduled_processing_time =
+                    Instant::now().add(initial_delay + slot_duration * position_in_queue as u32);
+                debug!(log, "Backfill work scheduled for processing");
+
+                self.backfill_delay_queue
+                    .insert_at(work_event, scheduled_processing_time);
+            }
```

Comment:
Checking this calculation it might be easier to store the batches in a `Vec` (probably a `SmallVec` is better suited here) and an `Interval` or a `Sleep` according to how important it is we keep certain alignment to the slot clock (probably a sleep), we do this in a couple of places already.

Examples: 
[`long_lived_subscriptions`](https://github.com/sigp/lighthouse/blob/fcfd02aeec435203269b03865e3ccc23e5f51e6d/beacon_node/network/src/subnet_service/attestation_subnets.rs#L76) paired with [`next_long_lived_subscription_event`](https://github.com/sigp/lighthouse/blob/fcfd02aeec435203269b03865e3ccc23e5f51e6d/beacon_node/network/src/subnet_service/attestation_subnets.rs#L113)
[`next_fork_update` and `next_fork_subscriptions`](https://github.com/sigp/lighthouse/blob/38514c07f222ff7783834c48cf5c0a6ee7f346d0/beacon_node/network/src/service.rs#L196-L198)

We would then poll the sleep, check the queued items and emit if not empty. I think this could be not just easier but more accurate than the current calculation

Reply:
I've implemented the approach you suggested, and I like it!  definitely agree it's a lot easier this way. 

I wonder if we need `SmallVec` (over `Option`) if there could ever only be one item in the `Vec`?

Reply:
I prefer the Smallvec option since it's still very low cost but it's future compatible, in the sense that if for some reason sync starts sending multiple batches in the future we won't run into problems. 

In this sense it should also make the code you write easier. You will most likely encounter the (right now impossible) code path in which you receive a batch and the option has something inside already. What do you do here? Using a smallvec you just push it and it will work just like a Vec (literally)

Reply:
make sense, thanks!

Reply:
We can probably stop polling once the backfill is complete - will look into this

Reply:
For an active check you can access the backfill state in the network globals. It's possible to have a watcher over the state to even avoid thet check

Reply:
Do you think this is still worth doing, given that it's no longer spawning a separate task?

Just thinking about the costs here - if we use the network globals, then we'll have to add it as a dependency to `ReprocessQueue` - not sure if this is really necessary.

With the watcher route, if I'm understanding it correctly, we'd have to create a `watch::Sender<EngineState>` over the  `backfill_state` in `NetworkGlobals`, and sends a message to this channel any time it's updated. And within the `ReprocessQueue` we would listen to this channel and flip a boolean flag (e.g. is_backfill_in_progress) when backfill completes, and then stop polling the `next_backfill_batch_event` `Sleep`?

Reply:
It wasn't a suggestion, just answering your question on different ways to handle this, since it can be as simple has making the sleep an option. We don't really care if backfill en enabled or not, or if it's completed etc, we just care about the presence of batches to be dealt with

Reply:
Thanks - went with making the sleep an option, that's definitely much simpler!